modules/gapi/include/opencv2/gapi/infer.hpp

```diff
@@ -174,41 +174,91 @@ struct InOutInfo
     std::vector<std::string> out_names;
 };
 
-/**
- * @{
- * @brief G-API object used to collect network inputs
- */
-class GAPI_EXPORTS_W_SIMPLE GInferInputs
+template <typename OutT>
+class GInferOutputsTyped
 {
-using Map = std::unordered_map<std::string, GMat>;
 public:
-    GAPI_WRAP GInferInputs();
-    GAPI_WRAP void setInput(const std::string& name, const cv::GMat& value);
-
-    cv::GMat& operator[](const std::string& name);
-    const Map& getBlobs() const;
+    GInferOutputsTyped() = default;
+    GInferOutputsTyped(std::shared_ptr<cv::GCall> call)
+        : m_priv(std::make_shared<Priv<OutT>>(std::move(call)))
+    {
+    }
 
+    OutT at(const std::string& name)
+    {
+        auto it = m_priv->blobs.find(name);
+        if (it == m_priv->blobs.end()) {
+            // FIXME: Avoid modifying GKernel
+            auto shape = cv::detail::GTypeTraits<OutT>::shape;
+            m_priv->call->kernel().outShapes.push_back(shape);
+            m_priv->call->kernel().outCtors.emplace_back(cv::detail::GObtainCtor<OutT>::get());
+            auto out_idx = static_cast<int>(m_priv->blobs.size());
+            it = m_priv->blobs.emplace(name,
+                    cv::detail::Yield<OutT>::yield(*(m_priv->call), out_idx)).first;
+            m_priv->info->out_names.push_back(name);
+        }
+        return it->second;
+    }
 private:
-    std::shared_ptr<Map> in_blobs;
+    template<typename T>
+    struct Priv
+    {
+        Priv(std::shared_ptr<cv::GCall> c)
+            : call(std::move(c)), info(cv::util::any_cast<InOutInfo>(&call->params()))
+        {
+        }
+
+        std::shared_ptr<cv::GCall> call;
+        InOutInfo* info = nullptr;
+        std::unordered_map<std::string, T> blobs;
+    };
+
+    std::shared_ptr<Priv<OutT>> m_priv;
 };
-/** @} */
 
-/**
- * @{
- * @brief G-API object used to collect network outputs
- */
-struct GAPI_EXPORTS_W_SIMPLE GInferOutputs
+using GInferOutputs     = GInferOutputsTyped<cv::GMat>;
```

Comment:
How to document these properly ?

Reply:
@dkurt Could you have a look, please ?

Reply:
https://pullrequest.opencv.org/buildbot/export/pr/19310/docs/d2/d75/namespacecv.html#a7f91c8222af32b19c81a31fffe4bcf8d Looks good